kernel/USER_GUIDE.md

```diff
@@ -0,0 +1,604 @@
+
+
+##### Method contextualizeFileReads() 
+
+This method associates the information of a to-be-read file to a connector-specific context object, called `FileReadContext`. Implementations can use this method to inject connector-specific information to optimize the file read. For example, if you want split reading a 256 MB Parquet file (say, `file1.parquet`) into 2 chunks of size 128 MB, then you can create two successive entries `fileReadContext1(file1.parquet)` and `fileReadContext1(file1.parquet)` where the context objects created by this method contains the byte ranges to read. This byte ranges will be used by the next method to do the necessary partial reads.
+
+**Requirements and guarantees: **Any implementation must adhere to the following guarantees.
+
+
+* The output iterator must maintain the same ordering as the input iterator. For example, if file1 is before file2 in the input iterator, then all the contexts associated with file1 must be before those of file2 in the output iterator.
+
+
+##### Method readParquetFiles() 
+
+This method takes as input information as `FileReadContext`s and returns the data in a series of columnar batches. The columns to be read from the Parquet file are defined by the physical schema, and the return batches must match that schema. To implement this method, you may have to first implement your own `<code>[ColumnarBatch](https://delta-io.github.io/delta/snapshot/kernel-api/java/api//io/delta/kernel/data/ColumnarBatch.html)`</code> and `<code>[ColumnVector](https://delta-io.github.io/delta/snapshot/kernel-api/java/api//io/delta/kernel/data/ColumnVector.html)`</code> which is used to represent the in-memory data generated from the Parquet files.
+
+When identifying the columns to read, note that there are multiple types of columns in the physical schema (represented as a StructType). 
+
+
+
+* Data columns: Columns that are expected to be read from the Parquet file. Based on the ``StructField`` object defining the column, there are two cases of mapping the column description to the right column in the file. 
+    * If the optional ``StructField.fieldId`` is defined, then read the column in the Parquet file which has the same field id.
+    * Otherwise, read the column in the Parquet file that matches the same name.
+* Metadata columns: These are special columns that must be populated using metadata about the Parquet file. To understand how to populate such a column, first match the column name against the set of standard metadata column name constants. For example, 
+    * ``StructField`.`ROW_INDEX_COLUMN_NAME``, then you have to a generate column vector populated with the actual index of each row in the Parquet file (that is, not indexed by the possibly subset of rows return after Parquet data skipping). See the docs for more details.
```

Comment:
This didn't render correctly:
![image](https://github.com/delta-io/delta/assets/79601771/0080340c-a4fb-401d-ba4e-b964ddcf79fa)
```suggestion
    * `` `StructField`.`ROW_INDEX_COLUMN_NAME` ``, then you have to a generate column vector populated with the actual index of each row in the Parquet file (that is, not indexed by the possibly subset of rows return after Parquet data skipping). See the docs for more details.
```

It apparently needs padding spaces to separate the "boundary" backticks from the "internal" ones that should render, ie `` `StructField`.`ROW_INDEX_COLUMN_NAME` ``. But anyway shouldn't those be single quotes rather than backticks? `'StructField'.'ROW_INDEX_COLUMN_NAME'`